### PR TITLE
Add mallctl for dumping last-N profiling records

### PR DESCRIPTION
**Pull request by yinan1048576**
I'm currently defining a `write_cb_packet_t` construct in `ctl.c` to represent the `[write_cb, cbopaque]` tuple passed in as input to the `mallctl()`.

Also added extensive unit tests.

### CODE REVIEW

File: src/ctl.c [hunk 1]
```diff
@@ -3549,3 +3551,34 @@ experimental_prof_recent_alloc_max_ctl(tsd_t *tsd, const size_t *mib,
...
+experimental_prof_recent_alloc_dump_ctl(tsd_t *tsd, const size_t *mib,
+    size_t miblen, void *oldp, size_t *oldlenp, void *newp, size_t newlen) {
+	int ret;
+
+	if (!(config_prof && opt_prof)) {
+		ret = ENOENT;
+		goto label_return;
+	}
+
+	assert(sizeof(write_cb_packet_t) == sizeof(size_t) * 2);
```
Comment (interwq) [inline src/ctl.c]: Why is this required?

File: test/unit/prof_recent.c [hunk 2]
```diff
@@ -381,6 +381,204 @@ TEST_END
...
+} confirm_record_t;
+
+#define ERROR "Dump output is wrong"
+
+static void confirm_record(const char *template,
+    const confirm_record_t *records, const size_t n_records) {
+	static const char *types[2] = {"alloc", "dalloc"};
+	static char buf[64];
+
+	assert_d_eq(memcmp(dump_out, template, strlen(template) - 2), 0, ERROR);
```
Comment (interwq) [inline test/unit/prof_recent.c]: nit: explain why the -2 offset.

File: test/unit/prof_recent.c [hunk 3]
```diff
@@ -381,6 +381,204 @@ TEST_END
...
+	for (record = records; record < records + n_records; ++record) {
+		if (record != records) {
+			assert_true(start < end, ERROR);
+			assert_c_eq(*start++, ',', ERROR);
+		}
+
+		assert_true(start < end, ERROR);
+		assert_c_eq(*start++, '{', ERROR);
+
+		{
```
Comment (interwq) [inline test/unit/prof_recent.c]: Abstract the three blocks below into a function?

File: test/unit/prof_recent.c [hunk 4]
```diff
@@ -381,6 +381,204 @@ TEST_END
...
+			malloc_snprintf(buf, sizeof(buf), "%s",
+			    record->released ? "true" : "false");
+			assert_d_eq(memcmp(start, buf, strlen(buf)), 0, ERROR);
+			start += strlen(buf);
+			assert_c_eq(*start++, ',', ERROR);
+		}
+
+		const char **type = types;
+		while (true) {
+			assert_true(start < end, ERROR);
```
Comment (interwq) [inline test/unit/prof_recent.c]: Similarly, the three blocks here.

File: src/ctl.c (hunk 1 above)
Comment (yinan1048576) [reply src/ctl.c]: It's similar to what we had for the memory utilization APIs. See https://github.com/jemalloc/jemalloc/blob/7014f81e172290466e1a28118b622519bbbed2b0/src/ctl.c#L3442.

The input structure is not public, so the caller would have to make up a `sizeof(size_t) * 2` input array, but internally I'm using this `write_cb_packet_t` structure, so I'd better assert that the two are equivalent.

**Comment by yinan1048576** [comment]
Decided to abstract at one layer lower.

File: src/ctl.c (hunk 1 above)
Comment (interwq) [reply src/ctl.c]: Here probably better to say `== sizeof(void *) * 2`

Comment (yinan1048576) [reply src/ctl.c]: Oh I see. Yes, you're right.

**Comment by yinan1048576** [comment]
Revised according to feedback.

**Comment by yinan1048576** [comment]
Noticed that my `ERROR` macro clashed with some internal macro in some environment. Changed it to `DUMP_ERROR`.
